### ISSUE TITLE
No big.Int support?

### ISSUE THREAD

**Issue by bonedaddy** [lib/pq#754]
While migrating models over to use `big.Int` instead of `int64` data types, I ran into the following issue
```
sql: Scan error on column index 6: unsupported Scan, storing driver.Value type int64 into type *big.Int
```

After doing some research, it appears that this postgresql driver, doesn't have support for `big.Int` data types? I ran into a similar issue when using arrays and slices of `int64` which required using the `Int64Array` type. Unfortunately, it appears that there is no type, or method for writing/reading `big.Int`?

For now it looks like I'll have to go with string types, and convert as needed before interacting with the database. 

For reference I'm using gorm to interface with the database, and am using this pq driver.

**Comment by mewwts** [comment]
Hey @postables I've had to do _something_ along the lines of this before
```golang
type BigInt big.Int

func (b *BigInt) Value() (driver.Value, nil) {
    if b != nil {
        return b.String(), nil
    }
    return nil, nil
}

func (b *BigInt) Scan(value interface{}) error {
    if value == nil {
        b = nil
    }
    switch t := value.(type) {
    case int64:
        b = (*BigInt)(value.(int64))
    default:
        return fmt.Errorf("Could not scan type %T into BigInt", t)
    }
}
```

**Comment by phuwn** [comment]
> Hey @postables I've had to do _something_ along the lines of this before
> 
> ```go
> type BigInt big.Int
> 
> func (b *BigInt) Value() (driver.Value, nil) {
>     if b != nil {
>         return b.String(), nil
>     }
>     return nil, nil
> }
> 
> func (b *BigInt) Scan(value interface{}) error {
>     if value == nil {
>         b = nil
>     }
>     switch t := value.(type) {
>     case int64:
>         b = (*BigInt)(value.(int64))
>     default:
>         return fmt.Errorf("Could not scan type %T into BigInt", t)
>     }
> }
> ```

Huge help for me. Thank you, wise man!

**Comment by tyliggity** [comment]
This still makes no sense.  Scanning an `int64` to a `big.Int` defeats the entire purpose.  The whole point of a `big.Int` is that it holds more than `int64` can.  So, by definition, this does not work.

**Comment by tyliggity** [comment]
It's worth noting that when dealing with very large numbers (i.e. crypto/Etherscan/etc.) the max value of Postgres BIGINT (which is the same as the max value for Go `int64`) may not work.  For example, you can only express about 9.2 ETH with numbers of that size due to requiring 18 digits at least.

So, even if you use BIGINT in Postgres, you literally cannot store any value higher than the Go `int64` type.  Therefore, there is no value in switching to `big.Int`.  The only way to store a value in Postgres higher than Go `int64` is therefore to use VARCHAR and then parse that to a `big.Int` in Go.

**Comment by tyliggity** [comment]
> Hey @postables I've had to do _something_ along the lines of this before
> 
> ```go
> type BigInt big.Int
> 
> func (b *BigInt) Value() (driver.Value, nil) {
>     if b != nil {
>         return b.String(), nil
>     }
>     return nil, nil
> }
> 
> func (b *BigInt) Scan(value interface{}) error {
>     if value == nil {
>         b = nil
>     }
>     switch t := value.(type) {
>     case int64:
>         b = (*BigInt)(value.(int64))
>     default:
>         return fmt.Errorf("Could not scan type %T into BigInt", t)
>     }
> }
> ```

This line `b = (*BigInt)(value.(int64))` automatically disqualifies any Postgres supported integer values larger than BIGINT.  Postgres's BIGINT type has the same max value as the Go `int64` type, yes.  However, Postgres supports much larger integer values via the DECIMAL and NUMERIC types.  In fact, you never need to use the Go `big.Int` type unless you are dealing with values beyond `int64`, which are also beyond the Postgres BIGINT type's capacity.

The code provided would not properly handle such values larger than `int64`.  In such cases, you would be scanning `[]byte` (or converted to a `string`) coming from the driver/DB and then use the `big.Int.SetString` method to assign that `string` to the `big.Int`.  Scanning an `int64` to `big.Int` is pretty useless.

**Comment by phuwn** [comment]
@tyliggity I think he just wants to give you the strategy to solve the problem, not the exact plan hehe.

In my case, I write these codes to handle `big.Float` and it works well for me.

```go
type BigFloat big.Float

func (b *BigFloat) Value() (driver.Value, error) {
	if b != nil {
		return (*big.Float)(b).String(), nil
	}
	return nil, nil
}

func (b *BigFloat) Scan(value interface{}) error {
	if value == nil {
		b = nil
	}

	switch t := value.(type) {
	case float64:
		(*big.Float)(b).SetFloat64(value.(float64))
	case []uint8:
		_, ok := (*big.Float)(b).SetString(string(value.([]uint8)))
		if !ok {
			return fmt.Errorf("failed to load value to []uint8: %v", value)
		}
	default:
		return fmt.Errorf("Could not scan type %T into BigFloat", t)
	}

	return nil
}
```

**Comment by tyliggity** [comment]
@phuwn It's not a matter of a "full solution", which is trivial, but rather about capacity and precision.  If you need `big.Int`, it means you need more than `int64`.  By casting the DB value to an `int64` first and then to `big.Int` you are defeating the entire purpose of using `big.Int` in the first place and you will get inaccurate values from the DB for anything above the max value for `int64`.  Your cast may even fail for overflow.

What you shared is even worse because the float type is very unreliable, which is why services that rely on precision of many significant digits (i.e. Etherscan) do not use floats at all.  Instead, they use huge integers.  So, in your code, when you cast the DB value to a float, it will not be an exact operation for many significant digits.  You won't have the same capacity issues as `int64` but you will have precision issues that make this solution unusable for many projects.

**Comment by phuwn** [comment]
@tyliggity Well yes, I didn't try to correct you or something. You are right about the huge diff between `big.Int` & `int64` usage. I just try to share my point of view on how I adopt that guy's idea. And yes, I haven't tried to parse the maximum number PostgreSQL can give, my solution can't be used on all projects. I'm only using `numeric` here in my current project hehe. And your idea of using `bigint` for floats on PostgreSQL is great, well text can't be used to do math stuff on queries so yeah I got nothing to complain about it, I think I'll do try it out on larger cases. So, relax.  Here, how about this for the solution of this `Issue`. What do you think? Or you can share your own solution with us. That would be great, it would help us a lot. Codes are way better than talks haha.

```go
type BigInt big.Int

func (b *BigInt) Value() (driver.Value, error) {
	if b != nil {
		return (*big.Int)(b).String(), nil
	}
	return nil, nil
}

func (b *BigInt) Scan(value interface{}) error {
	if value == nil {
		b = nil
	}

	switch t := value.(type) {
	case []uint8:
		_, ok := (*big.Int)(b).SetString(string(value.([]uint8)))
		if !ok {
			return fmt.Errorf("failed to load value to []uint8: %v", value)
		}
	default:
		return fmt.Errorf("Could not scan type %T into BigInt", t)
	}

	return nil
}
```

**Comment by tyliggity** [comment]
No worries man just trying to point out the issues I've run into myself.

The solution for reads is to just use DECIMAL/NUMERIC on the Postgres side, which brings the value into Go as a slice of bytes which you can then just convert to string and use the `big.Int.SetString` method.

The solution for writes is to provide the `big.Int` as a string via the `big.Int.String` method.  Postgres will then coerce the string into a DECIMAL/NUMERIC type for you.  Nowhere in this process does there appear any `int64`, decimal places, or floats.

**Comment by phucledien** [comment]
In my case, Im using numeric data type in postgres and just use string for model declaration
To init the big int instance just need to call `new(big.Int).SetString()`
```go
type Model struct {
    BigIntField string `gorm:"type:numeric"`
}
```
Hope it will help 😄

**Comment by arp242** [comment]
pq can't directly support big.Int, because database/sql just doesn't give the ability to do that. It only supports the basic primitives (string, int, etc.) and time.Time.

So we would have to add our own type in the pq package that implements the Value() and Scan() methods, like some of the snippets posted in earlier comments.

I'm not principally against that, but that probably implement full decimal support, instead of just bigint. There's a few packages that do that already; https://github.com/shopspring/decimal is popular (it lists some alternatives in the README).

And if you need to bring your own type, I don't really see the advantage of using `pq.Decimal` over `decimal.Decimal`, other than saving a dependency? If I look at e.g. github.com/shopspring/decimal then it's over 7,000 lines of code. pq is about 18,000 lines. So that's a lot. I think it's too much to maintain here.

So in short, closing as "not something that database/sql doesn't allow us to implement" and "use a decimal package or your own type".
